# Werkcollege-opdrachten Week 1.3

## Dependencies importeren

Kopieer in het codeblok hieronder van het vorige werkcollege de import-code voor de dependencies die het vaakst worden gebruikt om data in te lezen. Geef er ook de gebruikelijke aliassen aan.<br>
Zet eventuele warnings uit.

In [ ]:
import pandas as pd
import sqlite3
import warnings
warnings.simplefilter('ignore')

pandas versie: 3.0.1


Zet de volgende bestanden in een makkelijk te vinden map, als je dat nog niet gedaan hebt:
- go_sales.sqlite
- go_crm.sqlite
- go_staff.sqlite

## Databasetabellen inlezen

Kopieer in het codeblok hieronder van het vorige werkcollege de code om een connectie met het bestand go_sales.sqlite te maken.

In [ ]:
salesConn = sqlite3.connect('go_sales.sqlite')

Verbonden met database: go_sales.sqlite


Lees van de ingelezen go_sales-database te volgende tabellen in met behulp van "SELECT * FROM *tabel*".
- product
- product_type
- product_line
- sales_staff
- sales_branch
- retailer_site
- country
- order_header
- order_details
- returned_item
- return_reason

In [ ]:
tables = [
    'product',
    'product_type',
    'product_line',
    'sales_staff',
    'sales_branch',
    'retailer_site',
    'country',
    'order_header',
    'order_details',
    'returned_item',
    'return_reason'
]

dfs = {table: pd.read_sql(f"SELECT * FROM {table}", salesConn) for table in tables}

for table, df in dfs.items():
    print(f"{table}: {df.shape} rijen x kolommen")

product = dfs['product']
product_type = dfs['product_type']
product_line = dfs['product_line']
sales_staff = dfs['sales_staff']
sales_branch = dfs['sales_branch']
retailer_site = dfs['retailer_site']
country = dfs['country']
order_header = dfs['order_header']
order_details = dfs['order_details']
returned_item = dfs['returned_item']
return_reason = dfs['return_reason']

product: (115, 9) rijen x kolommen
product_type: (21, 3) rijen x kolommen
product_line: (5, 2) rijen x kolommen
sales_staff: (102, 10) rijen x kolommen
sales_branch: (28, 7) rijen x kolommen
retailer_site: (391, 9) rijen x kolommen
country: (21, 4) rijen x kolommen
order_header: (4784, 8) rijen x kolommen
order_details: (37757, 7) rijen x kolommen
returned_item: (619, 5) rijen x kolommen
return_reason: (5, 2) rijen x kolommen


Krijg je een "no such table" error? Dan heb je misschien met .connect() per ongeluk een leeg  databasebestand (.sqlite) aangemaakt.<br> <u>Let op:</u> lees eventueel de informatie uit het vorige werkcollege nog eens goed door.

Als je tijdens onderstaande opdrachten uit het oog verliest welke tabellen er allemaal zijn, kan je deze Pythoncode uitvoeren:

In [5]:
sql_query = "SELECT * FROM sqlite_master WHERE type='table';"
#Vul dit codeblok verder in
pd.read_sql(sql_query, ...)
#Op de puntjes hoort de connectie naar go_sales óf go_staff óf go_crm te staan.

AttributeError: 'ellipsis' object has no attribute 'cursor'

Let op! Voor alle onderstaande opdrachten mag je <u>alleen Python</u> gebruiken, <u>geen SQL!</u>

## Selecties op één tabel zonder functies

Geef een overzicht met daarin de producten en hun productiekosten waarvan de productiekosten lager dan 100 dollar en hoger dan 50 dollar ligt. (2 kolommen, 23 rijen)

In [ ]:
kost_kolom = 'PRODUCTION_COST'
naam_kolom = 'PRODUCT_NAME'

overzicht_product = product.loc[(product[kost_kolom] > 50) & (product[kost_kolom] < 100), [naam_kolom, kost_kolom]]

overzicht_product = overzicht_product.reset_index(drop=True)

print(overzicht_product)
print(f"Aantal rijen: {len(overzicht_product)}")

                       PRODUCT_NAME  PRODUCTION_COST
0         TrailChef Deluxe Cook Set            85.11
1            TrailChef Double Flame            75.00
2                   Hibernator Lite            60.00
3                        Hibernator            86.00
4   Hibernator Self - Inflating Mat            78.55
5               Hibernator Camp Cot            65.33
6      Canyon Mule Climber Backpack            62.50
7           Granite Climbing Helmet            52.86
8             Husky Harness Extreme            55.00
9                    Granite Hammer            57.13
10                        Polar Ice            73.33
11                     Polar Sports            80.00
12                       Polar Wave            60.00
13                    Polar Extreme            72.50
14                     Edge Extreme            80.00
15                        Seeker 35            79.19
16                        Seeker 50            92.58
17                   Seeker Extreme           

Geef een overzicht met daarin de producten en hun marge waarvan de marge lager dan 20 % of hoger dan 60 % ligt. (2 kolommen, 7 rijen) 

In [ ]:
naam_kolom = 'PRODUCT_NAME'
marge_kolom = 'MARGIN'

overzicht_marge = product.loc[(product[marge_kolom] < 0.20) | (product[marge_kolom] > 0.60), [naam_kolom, marge_kolom]]

overzicht_marge = overzicht_marge.reset_index(drop=True)

print(overzicht_marge)
print(f"Aantal rijen: {len(overzicht_marge)}")

                    PRODUCT_NAME  MARGIN
0            TrailChef Water Bag    0.33
1              TrailChef Canteen    0.23
2          TrailChef Kitchen Kit    0.28
3                  TrailChef Cup    0.28
4             TrailChef Cook Set    0.30
..                           ...     ...
110        Blue Steel Max Putter    0.55
111  Course Pro Golf and Tee Set    0.33
112          Course Pro Umbrella    0.50
113          Course Pro Golf Bag    0.60
114            Course Pro Gloves    0.50

[115 rows x 2 columns]
Aantal rijen: 115


Geef een overzicht met daarin de landen waar met francs wordt betaald. Sorteer de uitkomst op land.  (1 kolom, 3 rijen)

In [ ]:
print('country columns:', country.columns.tolist())

definition_col = 'CURRENCY_NAME'
land_col = 'COUNTRY'

franc_landen = country.loc[country[definition_col].str.contains('franc', case=False, na=False), [land_col]]
franc_landen = franc_landen.sort_values(by=land_col).reset_index(drop=True)

print(franc_landen)
print(f"Aantal rijen: {len(franc_landen)}")

country columns: ['COUNTRY_CODE', 'COUNTRY', 'LANGUAGE', 'CURRENCY_NAME']
       COUNTRY
0      Belgium
1       France
2  Switzerland
Aantal rijen: 3


Geef een overzicht met daarin de verschillende introductiedatums waarop producten met meer dan 50% marge worden geïntroduceerd (1 kolom, 7 rijen) 

In [ ]:
naam_kolom = 'INTRODUCTION_DATE'
marge_kolom = 'MARGIN'

introductiedatums = product.loc[product[marge_kolom] > 0.50, [naam_kolom]].drop_duplicates().sort_values(by=naam_kolom).reset_index(drop=True)

print(introductiedatums)
print(f"Aantal rijen: {len(introductiedatums)}")

  INTRODUCTION_DATE
0        1995-02-15
1        2000-10-26
2        2003-12-10
3        2003-12-15
4        2003-12-18
5        2003-12-27
6        2004-01-13
Aantal rijen: 7


Geef een overzicht met daarin het eerste adres en de stad van verkoopafdelingen waarvan zowel het tweede adres als de regio bekend is (2 kolommen, 7 rijen)

In [ ]:
print('sales_branch columns:', sales_branch.columns.tolist())

adres1_col = 'ADDRESS1' if 'ADDRESS1' in sales_branch.columns else 'ADDRESS_LINE_1'
adres2_col = 'ADDRESS2' if 'ADDRESS2' in sales_branch.columns else 'ADDRESS_LINE_2'
regio_col = 'REGION' if 'REGION' in sales_branch.columns else 'STATE_REGION'
stad_col = 'CITY' if 'CITY' in sales_branch.columns else 'TOWN'

resultaat = sales_branch.loc[
    sales_branch[adres2_col].notna() & (sales_branch[adres2_col] != '') &
    sales_branch[regio_col].notna() & (sales_branch[regio_col] != ''),
    [adres1_col, stad_col]
].drop_duplicates().reset_index(drop=True)

toont_resultaat = resultaat.head(7)

print(toont_resultaat)
print(f"Aantal rijen (filter resultaat): {len(resultaat)}")

sales_branch columns: ['SALES_BRANCH_CODE', 'ADDRESS1', 'ADDRESS2', 'CITY', 'REGION', 'POSTAL_ZONE', 'COUNTRY_CODE']
                           ADDRESS1              CITY
0               Singelgravenplein 4         Amsterdam
1  Prol. Paseo de la Reforma No. 51  Distrito Federal
2               202-2-3 Hyakunincho             Tokyo
3                     543-225 Asahi        Osaka City
4                  2315 Queen's Ave         Melbourne
5             Avenida Paulista, 333         São Paulo
6                    3 Albany Court        Birmingham
Aantal rijen (filter resultaat): 7


Geef een overzicht met daarin de landen waar met dollars (dollars of new dollar) wordt betaald. Sorteer de uitkomst op land. (1 kolom, 4 rijen) 

In [ ]:
valuta_col = 'CURRENCY_NAME'
land_col = 'COUNTRY'

dollar_landen = country.loc[country[valuta_col].str.contains('dollar', case=False, na=False), [land_col]]
dollar_landen = dollar_landen.drop_duplicates().sort_values(by=land_col).reset_index(drop=True)

print(dollar_landen)
print(f"Aantal rijen: {len(dollar_landen)}")

         COUNTRY
0      Australia
1         Canada
2         Taiwan
3  United States
Aantal rijen: 4


Geef een overzicht met daarin beide adressen en de stad van vestigingen van klanten waarvan de postcode begint met een ‘D’ (van duitsland). Filter op vestigingen die een tweede adres hebben. (3 kolommen, 2 rijen) 

In [30]:
print(retailer_site.columns.tolist())
print(retailer_site.head(10))

mask = retailer_site['POSTAL_ZONE'].astype(str).str.startswith('D', na=False)
print('D rows count', mask.sum())
print(retailer_site.loc[mask, ['ADDRESS1','ADDRESS2','CITY','POSTAL_ZONE','COUNTRY_CODE']].head(10))
print('Non-empty ADDRESS2 count among D-postal:', retailer_site.loc[mask & retailer_site['ADDRESS2'].notna(), 'ADDRESS2'].shape[0])

['RETAILER_SITE_CODE', 'RETAILER_CODE', 'ADDRESS1', 'ADDRESS2', 'CITY', 'REGION', 'POSTAL_ZONE', 'COUNTRY_CODE', 'ACTIVE_INDICATOR']
   RETAILER_SITE_CODE  RETAILER_CODE                          ADDRESS1  \
0                   1             89                1117 Franklin Blvd   
1                   2             89                   45, rue Atwater   
2                   3             89                  328 Hodgson Road   
3                   4             90                    2425, rue Peel   
4                   5             90                     34 8th Avenue   
5                   6             90                   45, rue Stanley   
6                   7             90               23 Louis Riel Drive   
7                   8             91               56 North Way Street   
8                   9             92                    601 Bay Street   
9                  10             92  Suite 1500, 802 Docklands Street   

  ADDRESS2         CITY            REGION POSTAL_ZON

## Selecties op één tabel met functies

Geef het totaal aantal producten dat is teruggebracht (1 waarde) 


Geef het aantal regio’s waarin verkoopafdelingen gevestigd zijn. (1 waarde)

Maak 3 variabelen:
- Een met de laagste
- Een met de hoogste
- Een met de gemiddelde (afgerond op 2 decimalen)

marge van producten (3 kolommen, 1 rij) 

Geef het aantal vestigingen van klanten waarvan het 2e adres niet bekend is (1 waarde)

Geef de gemiddelde kostprijs van de verkochte producten waarop korting (unit_sale_price < unit_price) is verleend (1 waarde) 

Geef een overzicht met daarin het aantal medewerkers per medewerkersfunctie (2 kolommen, 7 rijen) 

Geef een overzicht met daarin per telefoonnummer het aantal medewerkers dat op dat telefoonnummer bereikbaar is. Toon alleen de telefoonnummer waarop meer dan 4 medewerkers bereikbaar zijn. (2 kolommen, 10 rijen) 

## Selecties op meerdere tabellen zonder functies

Geef een overzicht met daarin het eerste adres en de stad van vestigingen van klanten uit ‘Netherlands’ (2 kolommen, 20 rijen) 

Geef een overzicht met daarin de productnamen die tot het producttype ‘Eyewear’ behoren. (1 kolom, 5 rijen) 

Geef een overzicht met daarin alle unieke eerste adressen van klantvestigingen en de voornaam en achternaam van de verkopers die ‘Branch Manager’ zijn en aan deze vestigingen hebben verkocht (3 kolommen, 1 rij) 

Geef een overzicht met daarin van de verkopers hun functie en indien zij iets hebben verkocht de datum waarop de verkoop heeft plaatsgevonden. Laat alleen de verschillende namen van de posities zien van de verkopers die het woord ‘Manager’ in hun positienaam hebben staan. (2 kolommen, 7 rijen) 

Geef een overzicht met daarin de verschillende namen van producten en bijbehorende namen van producttypen van de producten waarvoor ooit meer dan 750 stuks tegelijk verkocht zijn. (2 kolommen, 9 rijen) 

Geef een overzicht met daarin de productnamen waarvan ooit meer dan 40% korting is verleend. De formule voor korting is: (unit_price - unit_sale_price) / unit_price (1 kolom, 8 rijen) 

Geef een overzicht met daarin de retourreden van producten waarvan ooit meer dan 90% van de aangeschafte hoeveelheid is teruggebracht (return_quantity/quantity). (1 kolom, 3 rijen) 

## Selecties op meerdere tabellen met functies

Geef een overzicht met daarin per producttype het aantal producten die tot dat producttype behoren. (2 kolommen, 21 rijen) 

Geef een overzicht met daarin per land het aantal vestigingen van klanten die zich in dat land bevinden. (2 kolommen, 21 rijen) 

Geef een overzicht met daarin van de producten behorend tot het producttype ‘Cooking Gear’ per productnaam de totaal verkochte hoeveelheid en de gemiddelde verkoopprijs. Sorteer de uitkomst op totaal verkochte hoeveelheid. (6 kolommen, 10 rijen)

Geef een overzicht met daarin per land de naam van het land, de naam van de stad waar de verkoopafdeling is gevestigd (noem de kolomnaam in het overzicht ‘verkoper’) en het aantal steden waar zich klanten bevinden in dat land (noem de kolomnaam in het overzicht ‘klanten’) (3 kolommen, 29 rijen) 

## Pythonvertalingen van SUBSELECT en UNION met o.a. for-loops

Geef een overzicht met daarin de voornaam en de achternaam van de medewerkers die nog nooit wat hebben verkocht (2 kolommen, 25 rijen) 

Aantal teruggebrachte producten: 619


Geef een overzicht met daarin het aantal producten waarvan de marge lager is dan de gemiddelde marge van alle producten samen. Geef in het overzicht tevens aan wat de gemiddelde marge is van dit aantal producten waarvan de marge lager dan de gemiddelde marge van alle producten samen is. (1 kolom, 2 rijen) 

Geef een overzicht met daarin de namen van de producten die voor meer dan 500 (verkoopprijs) zijn verkocht maar nooit zijn teruggebracht. (1 kolom, 13 rijen) 

Geef een overzicht met daarin per (achternaam van) medewerker of hij/zij manager is of niet, door deze informatie toe te voegen als extra 'Ja/Nee'-kolom.<br>
Hint: gebruik een for-loop waarin je o.a. bepaalt of het woord 'Manager' in de functie (position_en) staat. (2 kolommen, 102 rijen).

Met de onderstaande code laat je Python het huidige jaar uitrekenen.

In [2]:
from datetime import date
date.today().year

2026

Met de onderstaande code selecteer je op een bepaald jaartal uit een datum.

In [3]:
from datetime import datetime

date_str = '16-8-2013'
date_format = '%d-%m-%Y'
date_obj = datetime.strptime(date_str, date_format)

date_obj.year

2013

Geef met behulp van bovenstaande hulpcode een overzicht met daarin op basis van het aantal jaar dat iemand in dienst is of een medewerker ‘kort in dienst’ (minder dan 30 jaar in dienst) of een ‘lang in dienst’ (groter dan gelijk aan 30 jaar in dienst) is. Geef daarbij per medewerker in een aparte kolom zowel ‘kort in dienst’ als ‘lang in dienst’ aan. Gebruik (wederom) een for-loop.<br>
(2 kolommen, 102 rijen) 